### PR TITLE
Processing Data In Models

### PR DESCRIPTION
**Pull request by HtetWaiYan7191**
### This pull request includes the following changes : 

Create data to operate CRUD operations in app 

#### Create custom methods for 

- Post
- User
- Comment
- Like
- Test all the methods in rails console 

Thanks for reviewing my codes 😁

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -2,4 +2,8 @@ class User < ApplicationRecord
   has_many :posts, foreign_key: 'author_id'
   has_many :likes
   has_many :comments
```
Comment (Munsa1) [inline app/models/user.rb]: - [x] Kindly add the foreign key as the `author_id` for the `Likes` and `Comments` as well as per requirement

File: app/models/comment.rb [hunk 2]
```diff
@@ -1,4 +1,8 @@
 class Comment < ApplicationRecord
   belongs_to :user
   belongs_to :post
```
Comment (Munsa1) [inline app/models/comment.rb]: - [x] Here we need to define an association between the Comment model and the User model as we did in the `Post` model. It indicates that a comment belongs to an author, and the author is an instance of the User model. Kindly fix this.

e.g
```
`belongs_to :author, class_name: 'User`
```

File: app/models/like.rb [hunk 3]
```diff
@@ -1,4 +1,8 @@
 class Like < ApplicationRecord
   belongs_to :user
   belongs_to :post
```
Comment (Munsa1) [inline app/models/like.rb]: - [x] Here we need to define an association between the Like model and the User model as we did in the `Post` model. It indicates that a like belongs to an author, and the author is an instance of the User model. Kindly fix this.

e.g
```
`belongs_to :author, class_name: 'User`
```

File: app/models/post.rb [hunk 4]
```diff
@@ -2,4 +2,12 @@ class Post < ApplicationRecord
...
   has_many :likes
   has_many :comments
+
+  def update_posts_counter(_count)
+    author.increment!(:post_counter)
+  end
+
+  def recent_comments(limit = 5)
+    comments.last(limit)
+  end
```
Comment (Munsa1) [inline app/models/post.rb]: - [x] Kindly consider making use of one of the [Rails Active Callbacks](https://guides.rubyonrails.org/active_record_callbacks.html) - `after_save` to call the `#update_posts_counter` method every time a new post has been saved to the database, then make the `update_post_counter` method private

e.g

```ruby
 after_save :update_post_counter

  private

  def update_posts_counter
    author.increment!(:posts_counter)
  end
```

File: app/models/comment.rb [hunk 5]
```diff
@@ -1,4 +1,8 @@
...
+
+  def update_comments_counter(count)
+    post.update(comment_counter: count)
+  end
 end
```
Comment (Munsa1) [inline app/models/comment.rb:13]: - [x] Kindly consider making use of one of the [Rails Active Callbacks](https://guides.rubyonrails.org/active_record_callbacks.html) - `after_save` to call the `#update_comments_counter` method every time a new comment has been saved to the database, then make the `update_comments_counter` method private

e.g

```ruby
 after_save :update_comments_counter

  private

  def update_comments_counter
    author.increment!(:comment_counter)
  end
```

File: app/models/like.rb [hunk 6]
```diff
@@ -1,4 +1,8 @@
...
+
+  def update_likes(count)
+    post.update(like_counter: count)
+  end
 end
```
Comment (Munsa1) [inline app/models/like.rb:13]: - [x] Kindly consider making use of one of the [Rails Active Callbacks](https://guides.rubyonrails.org/active_record_callbacks.html) - `after_save` to call the `#update_likes` method every time a new like has been saved to the database, then make the `update_likes` method private

e.g

```ruby
 after_save :update_likes

  private

  def update_likes
    author.increment!(like_counter)
  end
```

**Review by Munsa1 — CHANGES_REQUESTED** [review]
Hi @HtetWaiYan7191, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!
<FEEL FREE TO CHANGE THIS MESSAGE TO YOUR PERSONAL WORDS>

## Highlights
Has Separate Files For Each Model :heavy_check_mark: 
Correct Git Flow :heavy_check_mark: 
Descriptive Pr :heavy_check_mark: 

## Required Changes ♻️
_Check the comments under the review._
### :warning: Important :warning: 
- [ ] Kindly fix the broken link to your issues page under the `Contributing` section
image

- [ ] Kindly remove these commented-out lines from your readme file for a more cleaner and professional file.
image
image

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/models/like.rb (hunk 6 above)
Comment (HtetWaiYan7191) [reply app/models/like.rb:13]: Thank you so much for clear instructions sir 😁 .

**Review by thecodechaser — APPROVED** [review]
## STATUS: APPROVED 🟢

## Hi @HtetWaiYan7191👋, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## To Highlight 🎉
- Good job with processing data. ✔️
- No linter errors. :heavy_check_mark:
- Professional README file. :heavy_check_mark:
- Gitflow is followed. :heavy_check_mark:
- PR has a good title and summary. ✔️



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by thecodechaser** [comment]
@HtetWaiYan7191 Please fix the linters, They're failing now
You pushed the code after you submitted the project for review. Please make sure not to push any code after you submit the project for code review.

**Comment by HtetWaiYan7191** [comment]
Hello sir I did not push any codes while waiting for reviews .  I have no idea why the rubocop errors is happening . there is no error in my local.  As you can see there is no new commit I added . @thecodechaser .

**Comment by HtetWaiYan7191** [comment]
I will make sure to fix the linter error before mergin @thecodechaser . Thank you very much

**Pull request merged**